### PR TITLE
ci: caching in gh builds

### PR DESCRIPTION
**Pull request by rflechtner**
## fixes The Global Climate Crisis
Avoid unnecessary builds of the sdk to save the environment 🌳 🌍 🌳 
Builds the sdk once for each node version, zips the result, and uploads it as artefact. Each successive step (test, lint, integration test, run examples, etc.) just pulls the appropriate build, unzips, and continues from there. 

## How to test:
Just look at the checks it already ran.

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -30,6 +30,9 @@ jobs:
   build:
     runs-on: ubuntu-latest
 
+    env:
+      CI: true
```
Comment (tjwelde) [inline .github/workflows/tests.yml:34]: this should be set by github actions already automatically

**Review by tjwelde — COMMENTED** [review]
I like it

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -42,26 +45,78 @@ jobs:
...
       matrix:
         image: ['latest-develop', 'latest-rc']
 
     steps:
-      - uses: actions/checkout@v1
+      - uses: actions/download-artifact@v2
+        with:
+          name: build-12.x
+      - name: unzip
+        run: unzip build.zip -d .
```
Comment (tjwelde) [inline .github/workflows/tests.yml:119]: these steps could maybe be abstracted away. A custom action maybe?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (rflechtner) [reply .github/workflows/tests.yml:34]: ok, will remove it then. I had just moved it from the steps level to the job level.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (rflechtner) [reply .github/workflows/tests.yml:112]: there are zip/unzip actions out there, just felt like its overkill to replace a single command

Comment (rflechtner) [reply .github/workflows/tests.yml:112]: could just help with porting to different os

Comment (rflechtner) [reply .github/workflows/tests.yml:112]: we could use this https://github.com/TonyBogdanov/zip ; it uses the exact same tooling but launches an alpine linux container to do so :D

Comment (tjwelde) [reply .github/workflows/tests.yml:112]: no it's fine. as we talked directly, let's not overoptimize it

**Review by tjwelde — APPROVED** [review]

**Pull request merged**